File: models/Transaction.js
const mongoose = require('mongoose');

const TransactionSchema = new mongoose.Schema({
  user: { type: mongoose.Schema.ObjectId, ref: 'User', required: true },
  walletId: { type: mongoose.Schema.ObjectId, ref: 'Wallet' },
  tenant: { type: mongoose.Schema.ObjectId, ref: 'Tenant' },
  estate: { type: mongoose.Schema.ObjectId, ref: 'Estate' },
  amount: { type: Number, required: true, min: 0 },
  type: { type: String, enum: ['rent', 'utility', 'deposit', 'withdrawal', 'service_charge', 'caution_fee', 'legal_fee', 'maintenance', 'initial', 'bundle', 'other'], required: true },
  method: { type: String, enum: ['cash', 'transfer', 'card', 'bank', 'paystack', 'other'], default: 'transfer' },
  status: { type: String, enum: ['paid', 'pending', 'failed', 'completed'], default: 'paid' },
  reference: { type: String },
  periodMonth: { type: Number, min: 1, max: 12 },
  periodYear: { type: Number },
  notes: { type: String },
  description: { type: String },
  metadata: { type: mongoose.Schema.Types.Mixed },
  isActive: { type: Boolean, default: true },
  createdBy: { type: mongoose.Schema.ObjectId, ref: 'User', required: true },
  updatedBy: { type: mongoose.Schema.ObjectId, ref: 'User' }
}, { timestamps: true });

// Performance indexes for common query patterns
TransactionSchema.index({ user: 1, createdAt: -1 });
TransactionSchema.index({ tenant: 1, createdAt: -1 });
TransactionSchema.index({ estate: 1, createdAt: -1 });
TransactionSchema.index({ tenant: 1, status: 1, createdAt: -1 });
TransactionSchema.index({ status: 1, createdAt: -1 });
TransactionSchema.index({ estate: 1, status: 1, createdAt: -1 });

module.exports = mongoose.model('Transaction', TransactionSchema);

File: fastapi_app/models/transaction.py
from sqlalchemy import String, Boolean, DateTime, Float, Integer, Text
from sqlalchemy.orm import Mapped, mapped_column
from models.base import Base, Money, gen_uuid
from datetime import datetime
from utils.time_utils import utcnow


class Transaction(Base):
    __tablename__ = "transactions"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=gen_uuid)
    tenant: Mapped[str | None] = mapped_column(String(36), nullable=True, index=True)
    estate: Mapped[str | None] = mapped_column(String(36), nullable=True)
    user: Mapped[str | None] = mapped_column(String(36), nullable=True, index=True)
    wallet_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    amount: Mapped[float] = mapped_column(Money, default=0.0)
    type: Mapped[str] = mapped_column(String(100), default="rent")
    method: Mapped[str | None] = mapped_column(String(100), nullable=True)
    status: Mapped[str] = mapped_column(String(50), default="paid")
    reference: Mapped[str | None] = mapped_column(String(255), nullable=True)
    period_month: Mapped[int | None] = mapped_column(Integer, nullable=True)
    period_year: Mapped[int | None] = mapped_column(Integer, nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    created_by: Mapped[str | None] = mapped_column(String(36), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=utcnow, onupdate=utcnow)
